src/charm.py

```diff
@@ -0,0 +1,301 @@
+        """Returns a list of lightkube8s objects for all resources, rendered in charm context"""
+        return self._render_rbac() + self._render_crds() + self._render_controller()
+
+    def _check_deployed_resources(self):
+        """Check the status of all deployed resources, returning True if ok"""
+        expected_resources = self._render_all_resources()
+        found_resources = [None] * len(expected_resources)
+        for i, resource in enumerate(expected_resources):
+            self.logger.info(f"Checking for '{resource.metadata}'")
+            try:
+                found_resources[i] = get_k8s_obj(resource, self.lightkube_client)
+            except lightkube.core.exceptions.ApiError:
+                self.logger.info(
+                    f"Cannot find k8s object for metadata '{resource.metadata}'"
+                )
+        found_all_resources = all(found_resources)
+
+        statefulset_ok = validate_statefulsets(found_resources, self.logger)
+
+        return found_all_resources and statefulset_ok
+
+    @property
+    def manifest_file_root(self):
+        return self._manifests_file_root
+
+    @manifest_file_root.setter
+    def manifest_file_root(self, value):
+        self._manifests_file_root = Path(value)
+
+    def _get_manifest_files(self) -> list:
+        """Returns a list of all manifest files"""
+        return glob.glob(str(self.manifest_file_root / "*.yaml"))
+
+    @property
+    def lightkube_client(self):
+        if not self._lightkube_client:
+            self._lightkube_client = lightkube.Client()
+        return self._lightkube_client
+
+    @lightkube_client.setter
+    def lightkube_client(self, client):
+        self._lightkube_client = client
+
+
+def validate_statefulsets(objs, logger=None):
+    """Returns True if all StatefulSets in objs have the expected number of readyReplicas else False
+
+    Optionally emits a message to logger for any StatefulSets that do not have their desired number
+    of replicas
+    """
+    statefulset_not_ok = False
+    for obj in objs:
+        if isinstance(obj, StatefulSet):
+            readyReplicas = obj.status.readyReplicas
+            replicas_expected = obj.spec.replicas
+            if readyReplicas != replicas_expected:
+                statefulset_not_ok = True
+                if logger:
+                    logger.info(
+                        f"StatefulSet {obj.metadata.name} in namespace "
+                        f"{obj.metadata.namespace} has {readyReplicas} readyReplicas, "
+                        f"expected {replicas_expected}"
+                    )
+    return not statefulset_not_ok
+
+
+ALLOWED_IF_EXISTS = (None, "replace", "patch")
+
+
+def _validate_if_exists(if_exists):
+    if if_exists in ALLOWED_IF_EXISTS:
+        return if_exists
+    else:
+        raise ValueError(
+            f"Invalid value for if_exists '{if_exists}'.  Must be one of {ALLOWED_IF_EXISTS}"
+        )
+
+
+def create_all_lightkube_objects(
```

Comment:
I could go either way.  When I first wrote it, I wasn't passing `lightkube_client` or `logger` so it was completely static and felt like a helper.  Now I'm on the fence... 

Ideally though, I'd like to move this or something like it to either `lightkube` or maybe a generic helpers package we have as something like this would be useful in many charms

Reply:
I see, and I agree this could be used by many charms, though with [this](https://github.com/gtsystem/lightkube/issues/16), these lines would be reduced to just a couple.

Reply:
Lets leave it like it is right now in this codebase and consolidate once we agree on a library or a common solution.